dbms/src/Storages/Transaction/LearnerRead.cpp

```diff
@@ -208,29 +208,40 @@ LearnerReadSnapshot doLearnerRead(
 
         std::vector<kvrpcpb::ReadIndexRequest> batch_read_index_req;
         batch_read_index_req.reserve(ori_batch_region_size);
+        size_t stale_read_count = 0;
 
         {
             // If using `std::numeric_limits<uint64_t>::max()`, set `start-ts` 0 to get the latest index but let read-index-worker do not record as history.
             auto read_index_tso = mvcc_query_info->read_tso == std::numeric_limits<uint64_t>::max() ? 0 : mvcc_query_info->read_tso;
-
+            RegionTable & region_table = tmt.getRegionTable();
             for (size_t region_idx = region_begin_idx; region_idx < region_end_idx; ++region_idx)
             {
                 const auto & region_to_query = regions_info[region_idx];
                 const RegionID region_id = region_to_query.region_id;
-                if (auto ori_read_index = mvcc_query_info.getReadIndexRes(region_id); ori_read_index)
+                UInt64 physical_tso = read_index_tso >> TsoPhysicalShiftBits;
+                bool can_stale_read = physical_tso < region_table.getSelfSafeTS(region_id);
+                if (!can_stale_read)
                 {
-                    auto resp = kvrpcpb::ReadIndexResponse();
-                    resp.set_read_index(ori_read_index);
-                    batch_read_index_result.emplace(region_id, std::move(resp));
+                    if (auto ori_read_index = mvcc_query_info.getReadIndexRes(region_id); ori_read_index)
+                    {
+                        auto resp = kvrpcpb::ReadIndexResponse();
+                        resp.set_read_index(ori_read_index);
+                        batch_read_index_result.emplace(region_id, std::move(resp));
+                    }
+                    else
+                    {
+                        auto & region = regions_snapshot.find(region_id)->second;
+                        batch_read_index_req.emplace_back(GenRegionReadIndexReq(*region, read_index_tso));
+                    }
                 }
                 else
                 {
-                    auto & region = regions_snapshot.find(region_id)->second;
-                    batch_read_index_req.emplace_back(GenRegionReadIndexReq(*region, read_index_tso));
+                    batch_read_index_result.emplace(region_id, kvrpcpb::ReadIndexResponse());
```

Comment:
This will construct a default `ReadIndexResponse`, I'm not sure if there is some guarantee that read_index will always be `0`.

Reply:
> This will construct a default `ReadIndexResponse`, I'm not sure if there is some guarantee that read_index will always be `0`.

[For numeric types in proto3, the default value is zero.](https://developers.google.com/protocol-buffers/docs/proto3#default)